src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/MultiphasePoromechanicsReservoirFVM.hpp

```diff
@@ -0,0 +1,156 @@
+ */
+
+#ifndef GEOSX_LINEARALGEBRA_INTERFACES_HYPREMGRMULTIPHASEPOROMECHANICSRESERVOIRFVM_HPP_
+#define GEOSX_LINEARALGEBRA_INTERFACES_HYPREMGRMULTIPHASEPOROMECHANICSRESERVOIRFVM_HPP_
+
+#include "linearAlgebra/interfaces/hypre/HypreMGR.hpp"
+
+namespace geosx
+{
+
+namespace hypre
+{
+
+namespace mgr
+{
+
+/**
+ * @brief MultiphasePoromechanicsReservoirFVM strategy.
+ *
+ * Labels description stored in point_marker_array
+ *   - dofLabel: 0                = nodal displacement, x-component
+ *   - dofLabel: 1                = nodal displacement, y-component
+ *   - dofLabel: 2                = nodal displacement, z-component
+ *   - dofLabel: 3                = pressure
+ *   - dofLabel: 4                = density
+ *             ...                = densities
+ *   - dofLabel: numResLabels + 2 = density
+ *   - dofLabel: numResLabels + 3 = well density
+ *             ...                = well densities
+ *   - dofLabel: numResLabels + numWellLabels + 1 = well density
+ *   - dofLabel: numResLabels + numWellLabels + 2 = well rate
+ *
+ * 4-level MGR reduction strategy
+ *   - 1st level: eliminate displacements (0,1,2)
+ *   - 2nd level: eliminate the well block
+ *   - 3nd level: eliminate the reservoir density associated with the volume constraint (numLabels - 1)
+ *   - 4nd level: eliminate the other reservoir densities
+ *   - The coarse grid is solved with BoomerAMG.
+ *
+ */
+class MultiphasePoromechanicsReservoirFVM : public MGRStrategyBase< 4 >
+{
+public:
+  /**
+   * @brief Constructor.
+   * @param numComponentsPerField array with number of components for each field
+   */
+  explicit MultiphasePoromechanicsReservoirFVM( arrayView1d< int const > const & numComponentsPerField )
+    : MGRStrategyBase( LvArray::integerConversion< HYPRE_Int >( numComponentsPerField[0] + numComponentsPerField[1] + numComponentsPerField[2] ) )
+  {
+    // Level 0: eliminate displacement degrees of freedom
+    m_labels[0].resize( m_numBlocks - 3 );
+    std::iota( m_labels[0].begin(), m_labels[0].end(), 3 );
+
+    HYPRE_Int const numResLabels = LvArray::integerConversion< HYPRE_Int >( numComponentsPerField[1] );
+
+    // Level 1: eliminate the well block
+    m_labels[1].resize( numResLabels );
+    std::iota( m_labels[1].begin(), m_labels[1].end(), 3 );
+    // Level 2: eliminate reservoir last density which corresponds to the volume constraint equation
+    m_labels[2].resize( numResLabels - 1 );
+    std::iota( m_labels[2].begin(), m_labels[2].end(), 3 );
+    // Level 3: eliminate the remaining reservoir densities
+    m_labels[3].push_back( 3 );
+
+    setupLabels();
+
+    // Level 0
+    m_levelFRelaxMethod[0]     = MGRFRelaxationMethod::amgVCycle;
+    m_levelInterpType[0]       = MGRInterpolationType::jacobi;
+    m_levelRestrictType[0]     = MGRRestrictionType::injection;
+    m_levelCoarseGridMethod[0] = MGRCoarseGridMethod::nonGalerkin;
+
+    // Level 1
+    m_levelFRelaxMethod[1]     = MGRFRelaxationMethod::singleLevel;
```

Comment:
This means `Jacobi` for the well block. For small well blocks I'm not surprised it can work. I assume the plan is still to try Gaussian elimination as the well problem becomes larger.

Reply:
Yes, it is still the plan. Daniel has prepared a branch for me, it is really on me to try it on cases with wells now. I will try to do that soon.